### PR TITLE
GA to Lint + Format MegaBlocks

### PR DESCRIPTION
**Pull request by eitanturok**
This PR lints and formats MegaBlocks. Specifically, we:
* Create a `.pre-commit-config.yaml` with all of the same hooks as composer except for `pyright` (will be done in a subsequent PR).
* Formatted + Linted all code in MegaBlocks by running the `pre-commit` hooks. Changes include
	* 	adding a license
	* sorting imports
	* removing unusued imports
	* following style conventions in code
	*  following style conventions in docstring
	* lint all python code
	* lint all yaml code
* Added a Github Action to automatically perform code formatting + linting

Note: I made all the `.sh` files in the `exp` directory executable as per the `check-executables-have-shebangs` hook.

### CODE REVIEW

File: .github/workflows/code-quality.yaml [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+    working-directory: .
+jobs:
+  code-quality:
+    runs-on: ubuntu-latest  # todo: switch to linux-ubuntu-latest later
+    timeout-minutes: 30
+    strategy:
+      matrix:
+        python_version:
+        - "3.9"
+        - "3.10"
```
Comment (b-chu) [inline .github/workflows/code-quality.yaml:28]: pretty sure this is inconsistent

File: .github/workflows/code-quality.yaml [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+        pip_deps:
+        - "[dev]"
+    steps:
+    - uses: actions/checkout@v3
+    - name: Get composite run steps repository
+      uses: actions/checkout@v3
+      with:
+        repository: mosaicml/ci-testing
+        ref: v0.0.9
```
Comment (b-chu) [inline .github/workflows/code-quality.yaml:37]: outdated

File: .github/workflows/pr-gpu.yaml [hunk 3]
```diff
@@ -38,7 +38,7 @@ jobs:
         gpu_num: ${{ matrix.gpu_num }}
         git_repo: databricks/megablocks
         pip_deps: "[all,testing]"
-        pytest_command: "coverage run -m pytest tests" # todo: remove tests from pytest tests when we delete all tests outside of MegaBlocks repo
+        pytest_command: "coverage run -m pytest tests"  # todo: remove `tests` when delete tests outside megablocks dir
```
Comment (b-chu) [inline .github/workflows/pr-gpu.yaml:41]: why this change?

File: .github/workflows/pr-gpu.yaml [hunk 4]
```diff
@@ -15,7 +15,7 @@ concurrency:
 jobs:
   pytest-gpu:
     name: ${{ matrix.name }}
-    runs-on: ubuntu-latest # todo: switch to linux-ubuntu-latest later
+    runs-on: ubuntu-latest  # todo: switch to linux-ubuntu-latest later
```
Comment (b-chu) [inline .github/workflows/pr-gpu.yaml:18]: TODO

File: .pre-commit-config.yaml
Comment (b-chu) [inline .pre-commit-config.yaml:1]: Is this different from composer? How?

File: .yamllint.yaml
Comment (b-chu) [inline .yamllint.yaml:1]: Where'd you get this from?

File: megablocks/layers/__init__.py
Comment (b-chu) [inline megablocks/layers/__init__.py:1]: This doesn't match `megablocks/__init__.py`

File: pyproject.toml
Comment (b-chu) [inline pyproject.toml:1]: Where is this from?

**Review by b-chu — COMMENTED** [review]
Thanks for the PR!
- You have a bunch of empty file changes, remove those from this PR.
- Why don't you have pyright in this PR? There's some things like `__init__` inconsistencies that aren't getting caught.

File: pyproject.toml
Comment (eitanturok) [reply pyproject.toml:1]: This is taken from Composer.

**Comment by eitanturok** [comment]
> Thanks for the PR!
> 
> * You have a bunch of empty file changes, remove those from this PR.
> * Why don't you have pyright in this PR? There's some things like `__init__` inconsistencies that aren't getting caught.


The empty file changes are not actually empty changes. I made all of these files executable so the check-executables-have-shebangs hook would pass. It is kind of weird how this change appears in github. LMK if you think we should handle these changes differently.

Pyright raises 1.5k errors. I don't think we should add pyright in this PR becasue 1) these errors will take me a while to fix and 2) I think it is better to create smaller, more modular PRs. What do you think about this? cc-ing @mihir-db on this.

File: .github/workflows/code-quality.yaml (hunk 1 above)
Comment (eitanturok) [reply .github/workflows/code-quality.yaml:28]: [composer](https://github.com/mosaicml/composer/blob/358fbeb0c961eb206a183305554c2fda38488232/.github/workflows/code-quality.yaml#L24) tests against python 3.9, 3.10, 3.11. I'll add tests for 3.11. Is this what you meant about consistency?

File: .github/workflows/pr-gpu.yaml (hunk 3 above)
Comment (eitanturok) [reply .github/workflows/pr-gpu.yaml:41]: My initial comment was too long and the yaml linter complained.

File: .yamllint.yaml
Comment (eitanturok) [reply .yamllint.yaml:1]: Composer.

File: .pre-commit-config.yaml
Comment (eitanturok) [reply .pre-commit-config.yaml:1]: This is exactly the same as composer's `.pre-commit-config.yaml` except that this does not include pyright at the moment. Everything else is the same.

File: .github/workflows/pr-gpu.yaml (hunk 4 above)
Comment (eitanturok) [reply .github/workflows/pr-gpu.yaml:18]: I just turned all "todo"s  into "TODO"s.

File: .github/workflows/code-quality.yaml (hunk 2 above)
Comment (eitanturok) [reply .github/workflows/code-quality.yaml]: Just updated to `v0.1.0`

**Comment by eitanturok** [comment]
I addressed all of your comments @b-chu except for taking care of fixing the `__init__.py` files in MegaBlocks.

File: .github/workflows/code-quality.yaml (hunk 1 above)
Comment (b-chu) [reply .github/workflows/code-quality.yaml:28]: Depends on what python versions you want to support. @mihir-db ? I meant your gpu test was 3.11 only and this had 3.9 and 3.10 only.

File: .github/workflows/code-quality.yaml (hunk 2 above)
Comment (b-chu) [reply .github/workflows/code-quality.yaml]: outdated again

File: .github/workflows/pr-gpu.yaml (hunk 4 above)
Comment (b-chu) [reply .github/workflows/pr-gpu.yaml:19]: why can't you switch to linux-ubuntu-latest?

**Review by b-chu — CHANGES_REQUESTED** [review]
Fix the init files. Names should definitely match

File: .github/workflows/code-quality.yaml (hunk 2 above)
Comment (eitanturok) [reply .github/workflows/code-quality.yaml]: updated.

File: .github/workflows/pr-gpu.yaml (hunk 3 above)
Comment (eitanturok) [reply .github/workflows/pr-gpu.yaml]: I just fixed the comment and made it clearer. LMK if it is good now.

File: .github/workflows/pr-gpu.yaml (hunk 4 above)
Comment (eitanturok) [reply .github/workflows/pr-gpu.yaml:19]: linux-ubuntu-latest is a custom runner that we at Databricks have. We have not fully set it up yet for this repo so we are using ubuntu-latest right now. @mihir-db previously gave the okay to use this runner.

File: megablocks/layers/__init__.py
Comment (eitanturok) [reply megablocks/layers/__init__.py:1]: I made it now match.

**Comment by eitanturok** [comment]
I also added a contributing guide, style guide, and Pull_REQUEST template, all initially taken from composer.

File: .github/workflows/code-quality.yaml (hunk 1 above)
Comment (eitanturok) [reply .github/workflows/code-quality.yaml:28]: Composer supports python 3.9, 3.10, 3.11 for their code-quality tests. We have this here.

Composer supports python 3.9 or 3.11 for their gpu-pr tests. In their yaml, the [container](https://github.com/mosaicml/composer/blob/a15b18ce18db6df4cff4e10251adcb9d5c5845db/.github/workflows/pr-gpu.yaml#L17) has python 3.11 but the [python-version](https://github.com/mosaicml/composer/blob/a15b18ce18db6df4cff4e10251adcb9d5c5845db/.github/workflows/pr-gpu.yaml#L32C7-L32C21) is 3.9. I chose to use python 3.11 because it is more up to date.

I do think it may be a bit repetitive/not the best use of GPU resources to run the gpu tests on multiple versions of python. However, I'm happy to add these tests if they are deemed necessary.

Comment (mvpatel2000) [reply .github/workflows/code-quality.yaml:28]: GPU tests should only be one python version whereas codeql is rotating 3 as we have in Composer. I think this is correct.

File: .github/workflows/pr-gpu.yaml [hunk 5]
```diff
@@ -16,7 +16,7 @@ jobs:
   pytest-gpu:
     name: ${{ matrix.name }}
     if: github.repository_owner == 'databricks'
-    runs-on: ubuntu-latest # todo: switch to linux-ubuntu-latest later
+    runs-on: ubuntu-latest  # todo: switch to linux-ubuntu-latest later
```
Comment (mvpatel2000) [inline .github/workflows/pr-gpu.yaml:19]:
```suggestion
    runs-on: ubuntu-latest  # TODO: switch to linux-ubuntu-latest later
```
nit consistent formatting

Comment (mvpatel2000) [reply .github/workflows/pr-gpu.yaml:19]: Can you file a JIRA for this

**Review by mvpatel2000 — COMMENTED** [review]
can you change license to `Databricks`? Otherwise LGTM, would also like @b-chu to approve since large PR

File: .github/workflows/pr-gpu.yaml (hunk 5 above)
Comment (eitanturok) [reply .github/workflows/pr-gpu.yaml]: Done.

File: megablocks/layers/__init__.py
Comment (eitanturok) [reply megablocks/layers/__init__.py:1]: I took a look at the files where megablocks is used in llm-foundry and put the functions / classes used in the top level __init__.py.

File: .github/workflows/code-quality.yaml (hunk 1 above)
Comment (eitanturok) [reply .github/workflows/code-quality.yaml:28]: As per in-person discussion, we are only supporting python 3.11 and GA's no longer running on 3.9 or 3.10.

**Review by b-chu — CHANGES_REQUESTED** [review]
Fix all your file headers

File: .github/workflows/pr-gpu.yaml (hunk 3 above)
Comment (eitanturok) [reply .github/workflows/pr-gpu.yaml]: put on a new line.

**Comment by eitanturok** [comment]
@mihir-db I changed the license.

**Review by b-chu — APPROVED** [review]
Thanks for the PR!
